Issue 1273: Migration: SafetiBase

simplypixi opened the pull request:
This fixes #1273 

#### Description
- Migrated SafetiBase
- Removed SafetiBase code old code
- Partialy migrated ViewerService
- Migrated MultSelectService (Please, check it carefully. We've migrated this service, but didn't know how to test it.)

simplypixi commented:
@carmenfan It was a loooong fight. A lot of code was migrated to React in this PR, so probably you'll find some bugs.

carmenfan commented:
@simplypixi thanks! I just tried building it, i'm getting the following errors (which is the same error as travis):
![image](https://user-images.githubusercontent.com/11945337/51489033-07d64800-1d9f-11e9-92f3-711ebe8c5d60.png)

simplypixi commented:
:o
Fixed!

carmenfan commented on frontend/components/viewer/js/viewer.service.ts:
```diff
@@ -186,15 +186,13 @@ export class ViewerService {
 	}
 
-	public getCurrentViewpoint(params) {
+	public async getCurrentViewpoint({ account, model, promise }) {
```
I'm guessing this is only taking a promise because issues isn't changed?

When it's all changed over can we also make it to take 2 parameters instead of an object with 2 parameters?

simplypixi replied:
Yes and yes ;)

carmenfan commented:
Sorry guys I found another issue but I forgot to post it with the earlier review:

The filter search used to prompt on potential data (like jobs, status etc) but it no longer does it. Can we add it back in? (right is staging):

![1](https://user-images.githubusercontent.com/11945337/51540974-97d3ca80-1e4f-11e9-9c59-36ffa90ca919.gif)

simplypixi commented:
> Sorry guys I found another issue but I forgot to post it with the earlier review:
> 
> The filter search used to prompt on potential data (like jobs, status etc) but it no longer does it. Can we add it back in? (right is staging):
> 
> ![1](https://user-images.githubusercontent.com/11945337/51540974-97d3ca80-1e4f-11e9-9c59-36ffa90ca919.gif)

Done ;)

dleszcz commented:
Hi @carmenfan. 
Can I ask you to send us the response from backend GET **risks** method?
It can help us with these errors: 
- Pins are not being shown on load
- Clicking on risks cause an error (and probably related, but viewpoints and objects are nto being shown)

carmenfan commented:
@simplypixi :

> Sure! We use the same component, which was used to creating user managment' pages. Should we change it globally or only here? What do you prefer?

Maybe just here, Looking at the user management all the font is grey. So unless you fancy changing all that it's probably easier to just change it here 😆 

> Do you have the latest version of Risks? We've hidden this save button because of that each field is updated dynamically - we added the snackbar, which inform about that on each saving. If you want to have Risks and Issues working in the same way, we have to do this. After saving new risk/issue, in this place will display comment form. So the save button will be used to saving new comments.

I just tried the latest and I think its fine if you want to do it that way. But I'm not a big fan of text auto-saving like this. Can we make it so we have inline icons like views (unlike views, probably should be in the same line as the text box):

![1](https://user-images.githubusercontent.com/11945337/51600092-1cc6ee80-1ef9-11e9-8a63-e1ebc8150677.gif)

> Is it any reason, why the risk owner ('assigned_roles') is stored as an array? If not, maybe is a good moment to change it?

It's a compatibility thing with other software platforms, and also a wish to support multiple assignees in the future. It's probably best to keep it as is :)

@dleszcz :

> Can I ask you to send us the response from backend GET **risks** method?
> It can help us with these errors:

I tried the latest version and pins seems to be showing now. :) 

<details>
````json
[
   {
      "_id":"ebc8dea0-cbb1-11e8-a413-db67a6433de0",
      "rev_id":"923570be-3163-4b25-9cbd-cfac2bd3405e",
      "creator_role":"Carmen job",
      "name":"dfdsfsd",
      "associated_activity":"fsdfds",
      "viewpoint":{
         "right":[
            0.9931749701499939,
            -1.862645149230957e-9,
            -0.1166340559720993
         ],
         "up":[
            -0.046786822378635406,
            0.9160158634185791,
            -0.3984042704105377
         ],
         "position":[
            -32.5124397277832,
            9.315811157226562,
            1.1740694046020508
         ],
         "look_at":[
            -33.70145797729492,
            4.8514580726623535,
            -8.950791358947754
         ],
         "view_dir":[
            -0.1068386435508728,
            -0.40114209055900574,
            -0.9097639918327332
         ],
         "near":0.10000000149011612,
         "far":83.14714050292969,
         "fov":1.0471975803375244,
         "aspect_ratio":1.98142409324646,
         "clippingPlanes":[

         ],
         "highlighted_group_id":"eb98cdf0-cbb1-11e8-a413-db67a6433de0",
         "hideIfc":true,
         "hidden_group_id":"eb99b850-cbb1-11e8-a413-db67a6433de0",
         "screenshot":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/ebc8dea0-cbb1-11e8-a413-db67a6433de0/screenshot.png",
         "guid":"53350a53-92ef-4de9-ba84-4f32bb872177",
         "screenshotSmall":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/ebc8dea0-cbb1-11e8-a413-db67a6433de0/screenshotSmall.png"
      },
      "assigned_roles":[
         [
            "Awssome job"
         ]
      ],
      "category":"safety_public",
      "likelihood":2,
      "consequence":1,
      "level_of_risk":2,
      "mitigation_status":"agreed_partial",
      "position":[
         -31.870079040527344,
         2.7319846153259277,
         -6.965876579284668
      ],
      "norm":[
         0,
         0,
         0
      ],
      "owner":"carmen",
      "created":1539082566026,
      "desc":"Risk 1",
      "thumbnail":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/ebc8dea0-cbb1-11e8-a413-db67a6433de0/thumbnail.png",
      "mitigation_desc":"agreed_partial",
      "safetibase_id":"",
      "typePrefix":"Architectural",
      "modelCode":"ARCH",
      "account":"carmen",
      "model":"5da556e1-4f45-4b19-8564-f07447df1fe2"
   },
   {
      "_id":"06539740-ecd9-11e8-a41a-d713ac561baa",
      "rev_id":"923570be-3163-4b25-9cbd-cfac2bd3405e",
      "creator_role":"Carmen job",
      "name":"new risk",
      "associated_activity":"",
      "viewpoint":{
         "right":[
            0.5441068410873413,
            1.4901161193847656e-8,
            -0.8390159606933594
         ],
         "up":[
            -0.4813207983970642,
            0.8190844058990479,
            -0.3121393918991089
         ],
         "position":[
            21.304885864257812,
            31.481204986572266,
            9.208550453186035
         ],
         "look_at":[
            -15.225566864013672,
            0.9867669343948364,
            -14.481669425964355
         ],
         "view_dir":[
            -0.6872248649597168,
            -0.5736730098724365,
            -0.4456694722175598
         ],
         "near":0.21139883995056152,
         "far":105.69941711425781,
         "fov":1.0471975803375244,
         "aspect_ratio":1.98142409324646,
         "clippingPlanes":[
            {
               "normal":[
                  0,
                  0,
                  -1
               ],
               "distance":-19.243303298950195,
               "clipDirection":-1
            },
            {
               "normal":[
                  0,
                  0,
                  1
               ],
               "distance":66.10316467285156,
               "clipDirection":-1
            },
            {
               "normal":[
                  1,
                  0,
                  0
               ],
               "distance":0,
               "clipDirection":-1
            },
            {
               "normal":[
                  -1,
                  0,
                  0
               ],
               "distance":52.685001373291016,
               "clipDirection":-1
            },
            {
               "normal":[
                  0,
                  -1,
                  0
               ],
               "distance":14.754161834716797,
               "clipDirection":-1
            },
            {
               "normal":[
                  0,
                  1,
                  0
               ],
               "distance":0,
               "clipDirection":-1
            }
         ],
         "highlighted_group_id":"05eda980-ecd9-11e8-a41a-d713ac561baa",
         "hideIfc":true,
         "hidden_group_id":"05ee6cd0-ecd9-11e8-a41a-d713ac561baa",
         "screenshot":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/06539740-ecd9-11e8-a41a-d713ac561baa/screenshot.png",
         "guid":"0e403136-fe27-4f67-971d-2d4965c6a574",
         "screenshotSmall":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/06539740-ecd9-11e8-a41a-d713ac561baa/screenshotSmall.png"
      },
      "assigned_roles":[

      ],
      "category":"",
      "likelihood":0,
      "consequence":0,
      "level_of_risk":0,
      "mitigation_status":"",
      "owner":"carmen",
      "created":1542727749300,
      "desc":"dfsdfds",
      "thumbnail":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/06539740-ecd9-11e8-a41a-d713ac561baa/thumbnail.png",
      "mitigation_desc":"dsfsdffsdsd",
      "typePrefix":"Architectural",
      "modelCode":"ARCH",
      "account":"carmen",
      "model":"5da556e1-4f45-4b19-8564-f07447df1fe2"
   },
   {
      "_id":"ad41b050-1b14-11e9-9141-5dad5b0024d9",
      "rev_id":"923570be-3163-4b25-9cbd-cfac2bd3405e",
      "creator_role":"Carmen job",
      "name":"Test",
      "safetibase_id":"dsfsd",
      "associated_activity":"dsfsd",
      "desc":"sdfsfsd",
      "viewpoint":{
         "right":[
            0.7820074558258057,
            -1.4901161193847656e-8,
            0.6232690215110779
         ],
         "up":[
            0.10716722905635834,
            0.9851067662239075,
            -0.13446128368377686
         ],
         "position":[
            -71.58905029296875,
            18.935359954833984,
            33.778594970703125
         ],
         "look_at":[
            -25.962507247924805,
            6.1578826904296875,
            -23.468425750732422
         ],
         "view_dir":[
            0.6139865517616272,
            -0.1719437539577484,
            -0.7703608274459839
         ],
         "near":0.23443198204040527,
         "far":117.21598815917969,
         "fov":1.0471975803375244,
         "aspect_ratio":1.98142409324646,
         "clippingPlanes":[

         ],
         "highlighted_group_id":"",
         "hideIfc":true,
         "hidden_group_id":"ad1f5b40-1b14-11e9-9141-5dad5b0024d9",
         "screenshot":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/ad41b050-1b14-11e9-9141-5dad5b0024d9/screenshot.png",
         "guid":"00c08da4-b709-4b9d-b1fc-429ae3252231",
         "screenshotSmall":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/ad41b050-1b14-11e9-9141-5dad5b0024d9/screenshotSmall.png"
      },
      "assigned_roles":[
         "Another job"
      ],
      "category":"other",
      "likelihood":0,
      "consequence":0,
      "level_of_risk":0,
      "mitigation_status":"proposed",
      "owner":"carmen",
      "created":1547811123157,
      "thumbnail":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/ad41b050-1b14-11e9-9141-5dad5b0024d9/thumbnail.png",
      "mitigation_desc":"agreed_fully",
      "typePrefix":"Architectural",
      "modelCode":"ARCH",
      "account":"carmen",
      "model":"5da556e1-4f45-4b19-8564-f07447df1fe2"
   },
   {
      "_id":"7190c0d0-1e36-11e9-9a43-2b5bee8ba7d3",
      "name":"Untitled risk 4",
      "associated_activity":"fdsfsd",
      "assigned_roles":[
         "Carmen job"
      ],
      "likelihood":2,
      "consequence":0,
      "level_of_risk":0,
      "mitigation_status":"agreed_fully",
      "viewpoint":{
         "right":[
            0.10446840524673462,
            -2.9802322387695312e-8,
            0.9945283532142639
         ],
         "up":[
            0.4646623730659485,
            0.8841416835784912,
            -0.048809632658958435
         ],
         "position":[
            -92.59210205078125,
            34.10525894165039,
            -20.498348236083984
         ],
         "look_at":[
            -27.701786041259766,
            -0.3742636442184448,
            -27.314632415771484
         ],
         "view_dir":[
            0.8793039917945862,
            -0.46721887588500977,
            -0.0923648476600647
         ],
         "near":0.2304369956254959,
         "far":115.21849822998047,
         "fov":1.0471975803375244,
         "aspect_ratio":1.98142409324646,
         "clippingPlanes":[

         ],
         "highlighted_group_id":"",
         "hideIfc":true,
         "hidden_group_id":"715f9eb0-1e36-11e9-9a43-2b5bee8ba7d3",
         "screenshot":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/7190c0d0-1e36-11e9-9a43-2b5bee8ba7d3/screenshot.png",
         "guid":"2108a595-4894-4e73-ae9e-5c67d7b48946",
         "screenshotSmall":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/7190c0d0-1e36-11e9-9a43-2b5bee8ba7d3/screenshotSmall.png"
      },
      "owner":"carmen",
      "thumbnail":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/7190c0d0-1e36-11e9-9a43-2b5bee8ba7d3/thumbnail.png",
      "safetibase_id":"dsfsd",
      "mitigation_desc":"",
      "desc":"Another risk",
      "category":"environmental",
      "creator_role":"Carmen job",
      "position":[
         -26.064075469970703,
         8.331653594970703,
         -10.347912788391113
      ],
      "norm":[
         0,
         0,
         0
      ],
      "created":1548155479389,
      "rev_id":"923570be-3163-4b25-9cbd-cfac2bd3405e",
      "typePrefix":"Architectural",
      "modelCode":"ARCH",
      "account":"carmen",
      "model":"5da556e1-4f45-4b19-8564-f07447df1fe2"
   },
   {
      "_id":"a5833ee0-1e36-11e9-9a43-2b5bee8ba7d3",
      "name":"Untitled risk 5",
      "associated_activity":"",
      "assigned_roles":[

      ],
      "likelihood":0,
      "consequence":0,
      "level_of_risk":0,
      "mitigation_status":"",
      "viewpoint":{
         "right":[
            0.9761201739311218,
            7.450580596923828e-9,
            0.21723119914531708
         ],
         "up":[
            0.14894410967826843,
            0.727933406829834,
            -0.6692747473716736
         ],
         "position":[
            -39.371246337890625,
            50.22456359863281,
            25.12244415283203
         ],
         "look_at":[
            -27.701683044433594,
            -0.3744429349899292,
            -27.31430435180664
         ],
         "view_dir":[
            0.15812985599040985,
            -0.6856479048728943,
            -0.7105504274368286
         ],
         "near":0.21971458196640015,
         "far":109.85729217529297,
         "fov":1.0471975803375244,
         "aspect_ratio":1.98142409324646,
         "clippingPlanes":[

         ],
         "highlighted_group_id":"a5548dc0-1e36-11e9-9a43-2b5bee8ba7d3",
         "hideIfc":true,
         "hidden_group_id":"a5672b60-1e36-11e9-9a43-2b5bee8ba7d3",
         "screenshot":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/a5833ee0-1e36-11e9-9a43-2b5bee8ba7d3/screenshot.png",
         "guid":"30b5b5c0-a899-4698-88fc-79188ac98e1b",
         "screenshotSmall":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/a5833ee0-1e36-11e9-9a43-2b5bee8ba7d3/screenshotSmall.png"
      },
      "owner":"carmen",
      "thumbnail":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/a5833ee0-1e36-11e9-9a43-2b5bee8ba7d3/thumbnail.png",
      "creator_role":"Carmen job",
      "position":[
         -18.35476303100586,
         8.43885612487793,
         -50.20222473144531
      ],
      "norm":[
         0,
         0,
         0
      ],
      "created":1548155566542,
      "desc":"(No Description)",
      "rev_id":"923570be-3163-4b25-9cbd-cfac2bd3405e",
      "typePrefix":"Architectural",
      "modelCode":"ARCH",
      "account":"carmen",
      "model":"5da556e1-4f45-4b19-8564-f07447df1fe2"
   }
]
````
</details>

simplypixi commented:
> I just tried the latest and I think its fine if you want to do it that way. But I'm not a big fan of text auto-saving like this. Can we make it so we have inline icons like views (unlike views, probably should be in the same line as the text box):

I understand your point of view, but are you sure about inline icons? I mean, it's a good idea but it'll force an extra action to the user.
Hmm..., I have yet another idea. We can remove the dynamic update, handle saving form using the save button only and rearrange footer a bit (see the screenshot below).
We'll add another state for footer - "adding comments", which user can toggle using new button. It'll work similar to adding pins - while the option is enabled, the user can post comments instead of saving the form changes. What do you think about it?

![image](https://user-images.githubusercontent.com/5683490/51605840-78e93d00-1f10-11e9-9f99-212b46e39ea0.png)

> It's a compatibility thing with other software platforms, and also a wish to support multiple assignees in the future. It's probably best to keep it as is :)

Ok!

carmenfan commented:
Also, can we rename "Print" in the menu to "Create Report"? We're changing it on the Angular one for this release :)

simplypixi commented:
@carmenfan 
I've added icons to TextFields and changed the position of icons on the views list item. But, I had to remove autosuggest for `Associated Activity`. @paloknapo requested this feature if it's possible to implement. Currently, it is not, because of the new (wrapped) TextField component. I had to provide a few hacks to achieve confirmation action before saving the value, so TextField is currently incompatible with Autosuggest (because of it needs updated value before suggests something).

carmenfan commented:
@simplypixi I think that's fair, considering it's not a predefined list of values I'm not sure if it's appropriate anyway.

Is this ready for another look? :)

simplypixi commented:
> @simplypixi I think that's fair, considering it's not a predefined list of values I'm not sure if it's appropriate anyway.
> 
> Is this ready for another look? :)

Yes. Let me know if something should work in a different way. ;)

simplypixi commented on frontend/styles/colors.ts:
```diff
@@ -1,22 +1,59 @@
 	RED: '#FF0000'
 };
+
+export const PIN_COLORS = mapValues({
```
I think it doesn't make sense, because - in this case - COLORS constant was used to build PIN_COLORS. PIN_COLORS is related to COLORS but contains a different data type. eg.
```COLORS: '#ff0000',
PIN_COLORS: [255, 0, 0] ```
But you can still use it like COLORS eg. `PIN_COLORS.YELLOW`.

Let me know if it's not understandable to you. ;)

simplypixi commented on frontend/constants/risks.ts:
```diff
@@ -0,0 +1,148 @@
+	[RISK_LEVELS.REJECTED]: SyncProblem
+};
+
+export const RISK_LEVELS_COLORS = ['#008000', '#32cd32', '#fffacd', '#ff8c00', '#800000'];
```
Hmm, I'll have to create another constant but I think it's a good idea. 👍

simplypixi commented on frontend/constants/risks.ts:
```diff
@@ -0,0 +1,148 @@
+	[RISK_LEVELS.REJECTED]: SyncProblem
+};
+
+export const RISK_LEVELS_COLORS = ['#008000', '#32cd32', '#fffacd', '#ff8c00', '#800000'];
```
(duplicated) :)

simplypixi commented:
> 3. When loading risk from URL, it doesn’t appear to load saved viewpoint

@Charence  Are you sure it is a bug? Currently on staging if I load any risk from url the result is always the same:
![image](https://user-images.githubusercontent.com/5683490/51899585-9eb08f00-23b3-11e9-89e8-2968ed589828.png)

simplypixi commented:
> 1. Pin colour does not update before risk is saved
> 
> ![jan-28-2019 17-57-01](https://user-images.githubusercontent.com/1438332/51856139-d7eeed80-2326-11e9-8945-c669aa0a49c0.gif)
> 
> 2. Pin disappears if you click on pin immediately after it is saved
> 
> ![jan-28-2019 17-58-44](https://user-images.githubusercontent.com/1438332/51856156-e50bdc80-2326-11e9-9bd9-9081bac402f1.gif)

Fixed ;)

simplypixi commented:
@Charence I fixed bugs and left some comments. ;)

Charence commented:
> > 1. When loading risk from URL, it doesn’t appear to load saved viewpoint
> 
> @Charence Are you sure it is a bug? Currently on staging if I load any risk from url the result is always the same:

Interesting. It is a bug, but the bug also exists in master and staging. You have probably noticed that the camera viewpoint for a risk is not the same when the risk is loaded by following a URL compared to clicking on the risk from the risks panel? They should be the same.

The bug is most likely a result of a failed attempt to set the viewer's camera viewpoint since the viewer is still loading. Could you please see how to avoid this bug by waiting for the viewer to finish loading first? Issues should also have the same behaviour.

I remember at some point in the past, there was a timeout somewhere to load issues and risks after the viewer was loaded.

Let me know if this can be resolved in this branch or in another one.

simplypixi commented:
> Interesting. It is a bug, but the bug also exists in master and staging. You have probably noticed that the camera viewpoint for a risk is not the same when the risk is loaded by following a URL compared to clicking on the risk from the risks panel? They should be the same.

Yes, I noticed that. ;)

> The bug is most likely a result of a failed attempt to set the viewer's camera viewpoint since the viewer is still loading. Could you please see how to avoid this bug by waiting for the viewer to finish loading first? Issues should also have the same behaviour.

I think it's more complex, because currently setting camera position looks like this:
```
	public async setCamera(params) {
		await this.isViewerReady();
		return this.viewer.setCamera(
			params.position,
			params.view_dir,
			params.up,
			params.look_at,
			params.animate !== undefined ? params.animate : true,
			params.rollerCoasterMode,
			params.account,
			params.model
		);
	}
```
So it's waiting until the viewer is ready. :/
But I'll investigate the problem. Maybe the solution is easier to find then I think.

carmenfan commented:
I don't know why it's not showing here (I thinkt he revert of the revert is confusing github web interface) but there's conflicts with staging 😆

simplypixi commented:
> 2\. associated activity should be on the same line as the ID

@carmenfan Which browser did you use when you found this bug (fields position)? Because in Chrome everything looks fine:
![image](https://user-images.githubusercontent.com/5683490/52048028-a35b7b80-254a-11e9-87b2-065ac06c19ed.png)

simplypixi commented:
> * Risk tab should utilise the whole height where possible, unless it's sharing the space wiht another tab (left is staging)

Are you sure about that? Why e.g the views height isn't fit screen height? Maybe it's a better idea to fit the panel's height to the content if it is possible. What do you think?

carmenfan commented:
> > * Risk tab should utilise the whole height where possible, unless it's sharing the space wiht another tab (left is staging)
> 
> Are you sure about that? Why e.g the views height isn't fit screen height? Maybe it's a better idea to fit the panel's height to the content if it is possible. What do you think?

Yea I guess that's what I mean 😆 . The problem is I'm seeing the scroll bar unnecessarily (Also in the edit tab)

> @carmenfan Which browser did you use when you found this bug (fields position)? Because in Chrome everything looks fine:

I was using Chrome (windows 10). Looking at Charence's screenshot from the earlier review he also has the same problem. He was also using chrome (iOS Sierra 10.12.6)

simplypixi commented:
> I don't know why it's not showing here (I thinkt he revert of the revert is confusing github web interface) but there's conflicts with staging 😆
> 
> ![image](https://user-images.githubusercontent.com/11945337/51979599-71341600-2485-11e9-9a5a-4a6c465fec12.png)

It's because of the branch with shared components. First, we need to merge Risks there and then into the staging. So probably the conflicts will appear later. ;)

simplypixi commented:
> * Toggling on/off the pins seems to change the colour/position of the pins for some reason:

@carmenfan Could you send me your response (risks list) from the BE? I can't to reproduce this bug. :(

carmenfan commented:
Here's the reponse:
<details>

````
[
   {
      "_id":"ebc8dea0-cbb1-11e8-a413-db67a6433de0",
      "rev_id":"923570be-3163-4b25-9cbd-cfac2bd3405e",
      "creator_role":"Carmen job",
      "name":"dfdsfsd",
      "associated_activity":"fsdfds",
      "viewpoint":{
         "right":[
            0.9931749701499939,
            -1.862645149230957e-9,
            -0.1166340559720993
         ],
         "up":[
            -0.046786822378635406,
            0.9160158634185791,
            -0.3984042704105377
         ],
         "position":[
            -32.5124397277832,
            9.315811157226562,
            1.1740694046020508
         ],
         "look_at":[
            -33.70145797729492,
            4.8514580726623535,
            -8.950791358947754
         ],
         "view_dir":[
            -0.1068386435508728,
            -0.40114209055900574,
            -0.9097639918327332
         ],
         "near":0.10000000149011612,
         "far":83.14714050292969,
         "fov":1.0471975803375244,
         "aspect_ratio":1.98142409324646,
         "clippingPlanes":[

         ],
         "highlighted_group_id":"eb98cdf0-cbb1-11e8-a413-db67a6433de0",
         "hideIfc":true,
         "hidden_group_id":"eb99b850-cbb1-11e8-a413-db67a6433de0",
         "screenshot":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/ebc8dea0-cbb1-11e8-a413-db67a6433de0/screenshot.png",
         "guid":"53350a53-92ef-4de9-ba84-4f32bb872177",
         "screenshotSmall":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/ebc8dea0-cbb1-11e8-a413-db67a6433de0/screenshotSmall.png"
      },
      "assigned_roles":[
         "Another job"
      ],
      "category":"safety_public",
      "likelihood":2,
      "consequence":3,
      "level_of_risk":2,
      "mitigation_status":"proposed",
      "position":[
         -31.870079040527344,
         2.7319846153259277,
         -6.965876579284668
      ],
      "norm":[
         0,
         0,
         0
      ],
      "owner":"carmen",
      "created":1539082566026,
      "desc":"Risk 123dfjkdhsfjksd",
      "thumbnail":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/ebc8dea0-cbb1-11e8-a413-db67a6433de0/thumbnail.png",
      "mitigation_desc":"agreed_partialvfdvfdv",
      "safetibase_id":"",
      "typePrefix":"Architectural",
      "modelCode":"ARCH",
      "account":"carmen",
      "model":"5da556e1-4f45-4b19-8564-f07447df1fe2"
   },
   {
      "_id":"06539740-ecd9-11e8-a41a-d713ac561baa",
      "rev_id":"923570be-3163-4b25-9cbd-cfac2bd3405e",
      "creator_role":"Carmen job",
      "name":"new risk",
      "associated_activity":"",
      "viewpoint":{
         "right":[
            0.5441068410873413,
            1.4901161193847656e-8,
            -0.8390159606933594
         ],
         "up":[
            -0.4813207983970642,
            0.8190844058990479,
            -0.3121393918991089
         ],
         "position":[
            21.304885864257812,
            31.481204986572266,
            9.208550453186035
         ],
         "look_at":[
            -15.225566864013672,
            0.9867669343948364,
            -14.481669425964355
         ],
         "view_dir":[
            -0.6872248649597168,
            -0.5736730098724365,
            -0.4456694722175598
         ],
         "near":0.21139883995056152,
         "far":105.69941711425781,
         "fov":1.0471975803375244,
         "aspect_ratio":1.98142409324646,
         "clippingPlanes":[
            {
               "normal":[
                  0,
                  0,
                  -1
               ],
               "distance":-19.243303298950195,
               "clipDirection":-1
            },
            {
               "normal":[
                  0,
                  0,
                  1
               ],
               "distance":66.10316467285156,
               "clipDirection":-1
            },
            {
               "normal":[
                  1,
                  0,
                  0
               ],
               "distance":0,
               "clipDirection":-1
            },
            {
               "normal":[
                  -1,
                  0,
                  0
               ],
               "distance":52.685001373291016,
               "clipDirection":-1
            },
            {
               "normal":[
                  0,
                  -1,
                  0
               ],
               "distance":14.754161834716797,
               "clipDirection":-1
            },
            {
               "normal":[
                  0,
                  1,
                  0
               ],
               "distance":0,
               "clipDirection":-1
            }
         ],
         "highlighted_group_id":"05eda980-ecd9-11e8-a41a-d713ac561baa",
         "hideIfc":true,
         "hidden_group_id":"05ee6cd0-ecd9-11e8-a41a-d713ac561baa",
         "screenshot":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/06539740-ecd9-11e8-a41a-d713ac561baa/screenshot.png",
         "guid":"0e403136-fe27-4f67-971d-2d4965c6a574",
         "screenshotSmall":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/06539740-ecd9-11e8-a41a-d713ac561baa/screenshotSmall.png"
      },
      "assigned_roles":[

      ],
      "category":"",
      "likelihood":0,
      "consequence":0,
      "level_of_risk":0,
      "mitigation_status":"",
      "owner":"carmen",
      "created":1542727749300,
      "desc":"dfsdfds",
      "thumbnail":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/06539740-ecd9-11e8-a41a-d713ac561baa/thumbnail.png",
      "mitigation_desc":"dsfsdffsdsd",
      "typePrefix":"Architectural",
      "modelCode":"ARCH",
      "account":"carmen",
      "model":"5da556e1-4f45-4b19-8564-f07447df1fe2"
   },
   {
      "_id":"ad41b050-1b14-11e9-9141-5dad5b0024d9",
      "rev_id":"923570be-3163-4b25-9cbd-cfac2bd3405e",
      "creator_role":"Carmen job",
      "name":"Test",
      "safetibase_id":"dsfsd",
      "associated_activity":"dsfsd",
      "desc":"",
      "viewpoint":{
         "right":[
            0.7820074558258057,
            -1.4901161193847656e-8,
            0.6232690215110779
         ],
         "up":[
            0.10716722905635834,
            0.9851067662239075,
            -0.13446128368377686
         ],
         "position":[
            -71.58905029296875,
            18.935359954833984,
            33.778594970703125
         ],
         "look_at":[
            -25.962507247924805,
            6.1578826904296875,
            -23.468425750732422
         ],
         "view_dir":[
            0.6139865517616272,
            -0.1719437539577484,
            -0.7703608274459839
         ],
         "near":0.23443198204040527,
         "far":117.21598815917969,
         "fov":1.0471975803375244,
         "aspect_ratio":1.98142409324646,
         "clippingPlanes":[

         ],
         "highlighted_group_id":"",
         "hideIfc":true,
         "hidden_group_id":"ad1f5b40-1b14-11e9-9141-5dad5b0024d9",
         "screenshot":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/ad41b050-1b14-11e9-9141-5dad5b0024d9/screenshot.png",
         "guid":"00c08da4-b709-4b9d-b1fc-429ae3252231",
         "screenshotSmall":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/ad41b050-1b14-11e9-9141-5dad5b0024d9/screenshotSmall.png"
      },
      "assigned_roles":[
         "Another job"
      ],
      "category":"environmental",
      "likelihood":0,
      "consequence":2,
      "level_of_risk":0,
      "mitigation_status":"proposed",
      "owner":"carmen",
      "created":1547811123157,
      "thumbnail":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/ad41b050-1b14-11e9-9141-5dad5b0024d9/thumbnail.png",
      "mitigation_desc":"agreed_fully",
      "typePrefix":"Architectural",
      "modelCode":"ARCH",
      "account":"carmen",
      "model":"5da556e1-4f45-4b19-8564-f07447df1fe2"
   },
   {
      "_id":"7190c0d0-1e36-11e9-9a43-2b5bee8ba7d3",
      "name":"Untitled risk 4",
      "associated_activity":"fdsfsd",
      "assigned_roles":[
         "Carmen job"
      ],
      "likelihood":2,
      "consequence":0,
      "level_of_risk":0,
      "mitigation_status":"agreed_fully",
      "viewpoint":{
         "right":[
            0.10446840524673462,
            -2.9802322387695312e-8,
            0.9945283532142639
         ],
         "up":[
            0.4646623730659485,
            0.8841416835784912,
            -0.048809632658958435
         ],
         "position":[
            -92.59210205078125,
            34.10525894165039,
            -20.498348236083984
         ],
         "look_at":[
            -27.701786041259766,
            -0.3742636442184448,
            -27.314632415771484
         ],
         "view_dir":[
            0.8793039917945862,
            -0.46721887588500977,
            -0.0923648476600647
         ],
         "near":0.2304369956254959,
         "far":115.21849822998047,
         "fov":1.0471975803375244,
         "aspect_ratio":1.98142409324646,
         "clippingPlanes":[

         ],
         "highlighted_group_id":"",
         "hideIfc":true,
         "hidden_group_id":"715f9eb0-1e36-11e9-9a43-2b5bee8ba7d3",
         "screenshot":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/7190c0d0-1e36-11e9-9a43-2b5bee8ba7d3/screenshot.png",
         "guid":"2108a595-4894-4e73-ae9e-5c67d7b48946",
         "screenshotSmall":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/7190c0d0-1e36-11e9-9a43-2b5bee8ba7d3/screenshotSmall.png"
      },
      "owner":"carmen",
      "thumbnail":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/7190c0d0-1e36-11e9-9a43-2b5bee8ba7d3/thumbnail.png",
      "safetibase_id":"dsfsd",
      "mitigation_desc":"",
      "desc":"Another risk",
      "category":"environmental",
      "creator_role":"Carmen job",
      "position":[
         -26.064075469970703,
         8.331653594970703,
         -10.347912788391113
      ],
      "norm":[
         0,
         0,
         0
      ],
      "created":1548155479389,
      "rev_id":"923570be-3163-4b25-9cbd-cfac2bd3405e",
      "typePrefix":"Architectural",
      "modelCode":"ARCH",
      "account":"carmen",
      "model":"5da556e1-4f45-4b19-8564-f07447df1fe2"
   },
   {
      "_id":"a5833ee0-1e36-11e9-9a43-2b5bee8ba7d3",
      "name":"Untitled risk 5",
      "associated_activity":"",
      "assigned_roles":[
         "Awssome job"
      ],
      "likelihood":0,
      "consequence":0,
      "level_of_risk":0,
      "mitigation_status":"",
      "viewpoint":{
         "right":[
            0.9761201739311218,
            7.450580596923828e-9,
            0.21723119914531708
         ],
         "up":[
            0.14894410967826843,
            0.727933406829834,
            -0.6692747473716736
         ],
         "position":[
            -39.371246337890625,
            50.22456359863281,
            25.12244415283203
         ],
         "look_at":[
            -27.701683044433594,
            -0.3744429349899292,
            -27.31430435180664
         ],
         "view_dir":[
            0.15812985599040985,
            -0.6856479048728943,
            -0.7105504274368286
         ],
         "near":0.21971458196640015,
         "far":109.85729217529297,
         "fov":1.0471975803375244,
         "aspect_ratio":1.98142409324646,
         "clippingPlanes":[

         ],
         "highlighted_group_id":"a5548dc0-1e36-11e9-9a43-2b5bee8ba7d3",
         "hideIfc":true,
         "hidden_group_id":"a5672b60-1e36-11e9-9a43-2b5bee8ba7d3",
         "screenshot":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/a5833ee0-1e36-11e9-9a43-2b5bee8ba7d3/screenshot.png",
         "guid":"30b5b5c0-a899-4698-88fc-79188ac98e1b",
         "screenshotSmall":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/a5833ee0-1e36-11e9-9a43-2b5bee8ba7d3/screenshotSmall.png"
      },
      "owner":"carmen",
      "thumbnail":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/a5833ee0-1e36-11e9-9a43-2b5bee8ba7d3/thumbnail.png",
      "creator_role":"Carmen job",
      "position":[
         -18.35476303100586,
         8.43885612487793,
         -50.20222473144531
      ],
      "norm":[
         0,
         0,
         0
      ],
      "created":1548155566542,
      "desc":"(No Description)",
      "rev_id":"923570be-3163-4b25-9cbd-cfac2bd3405e",
      "safetibase_id":"",
      "category":"",
      "mitigation_desc":"",
      "typePrefix":"Architectural",
      "modelCode":"ARCH",
      "account":"carmen",
      "model":"5da556e1-4f45-4b19-8564-f07447df1fe2"
   },
   {
      "_id":"ede40020-248a-11e9-80ac-7fee36ce5db6",
      "name":"Untitled risk 6",
      "associated_activity":"",
      "assigned_roles":[

      ],
      "likelihood":4,
      "consequence":4,
      "level_of_risk":4,
      "mitigation_status":"",
      "viewpoint":{
         "right":[
            0.021494150161743164,
            -2.9802322387695312e-8,
            -0.9997689723968506
         ],
         "up":[
            -0.6295468807220459,
            0.7768446207046509,
            -0.013534724712371826
         ],
         "position":[
            -18.606822967529297,
            17.61624526977539,
            -9.1406831741333
         ],
         "look_at":[
            -33.154544830322266,
            5.821473121643066,
            -9.453446388244629
         ],
         "view_dir":[
            -0.776665210723877,
            -0.6296923160552979,
            -0.016697585582733154
         ],
         "near":0.10000000149011612,
         "far":62.669761657714844,
         "fov":1.0471975803375244,
         "aspect_ratio":1.98142409324646,
         "clippingPlanes":[

         ],
         "highlighted_group_id":"ecf95200-248a-11e9-80ac-7fee36ce5db6",
         "hideIfc":true,
         "hidden_group_id":"ed940b60-248a-11e9-80ac-7fee36ce5db6",
         "screenshot":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/ede40020-248a-11e9-80ac-7fee36ce5db6/screenshot.png",
         "guid":"9dbf45e1-1d67-485a-97cd-83e23799b691",
         "screenshotSmall":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/ede40020-248a-11e9-80ac-7fee36ce5db6/screenshotSmall.png"
      },
      "owner":"carmen",
      "thumbnail":"carmen/5da556e1-4f45-4b19-8564-f07447df1fe2/risks/ede40020-248a-11e9-80ac-7fee36ce5db6/thumbnail.png",
      "safetibase_id":"",
      "mitigation_desc":"",
      "category":"",
      "desc":"(No Description)",
      "creator_role":"Carmen job",
      "position":[
         -34.624267578125,
         7.302213668823242,
         -17.315853118896484
      ],
      "norm":[
         0,
         0,
         0
      ],
      "created":1548851472674,
      "rev_id":"923570be-3163-4b25-9cbd-cfac2bd3405e",
      "typePrefix":"Architectural",
      "modelCode":"ARCH",
      "account":"carmen",
      "model":"5da556e1-4f45-4b19-8564-f07447df1fe2"
   }
]
````

</details>

I think this may be to do with pins being dropped before the model started loading? It's vital that the model component has made a request to load the model (there's a promise for this on the viewer servicei believe)  before any interaction is made to the viewer, otherwise states of things may not be correct.. Here's another gif where one of the pin doesn't load initially.

![1](https://user-images.githubusercontent.com/11945337/52054744-5f21a880-2555-11e9-89b8-adf23487fd62.gif)

carmenfan commented:
Version 71.0.3578.98 (Official Build) (64-bit)

simplypixi commented:
@carmenfan 

> Version 71.0.3578.98 (Official Build) (64-bit)

Fixed ;)

> Risk tab should utilise the whole height where possible, unless it's sharing the space wiht another tab (left is staging)

And.... fixed. ;)  But the panel can "jumping" sometimes because of image loading. I mentioned about that before, but we plugged react's components into your current "height-setting" algorithm. I think the best effect we'll achieve after migration of all panels and that algorithm.

dleszcz commented:
@sanmont3drepo 
Can you check your requests again? 
Thanks :)

carmenfan commented:
@dleszcz 
There seems to be a lot of conflict with the base branch. If you can resolve that first then I can take a look

dleszcz commented:
> @dleszcz
> There seems to be a lot of conflict with the base branch. If you can resolve that first then I can take a look

Conflicts resolved :)

Charence commented:
> * I don't think the angular version does it, but I think the pins should update as the filter change (i.e. hidden issues needs to hide pin, shown issues need to show pins).

@dleszcz 
It does it. Pins visible on the model reflect the filtered list of risks.

Hiding filtered out pins will also solve the bug where risk details don't load when clicking on a pin for a risk that has been filtered out.

dleszcz commented:
@carmenfan Can you check your last requests again?

Charence commented:
> * Pressing on risk icons when the panel is hidden should still go to the viewpoint:
>   ![1](https://user-images.githubusercontent.com/11945337/52645013-d9e4af00-2ed7-11e9-900d-72b2c62f70eb.gif)
>   angular version:
>   ![1](https://user-images.githubusercontent.com/11945337/52645169-24662b80-2ed8-11e9-9f2a-eb1e232c28f8.gif)

@dleszcz @simplypixi 
In the past, clicking on a pin would also open the Issues or Safetibase panel if not already open. It looks like this functionality was lost and it is also missing from the version on staging.

simplypixi commented on frontend/globals/viewer.ts:
```diff
@@ -508,7 +513,14 @@ export class Viewer {
 
 		this.emit(Viewer.EVENT.START_LOADING);
```
I think it shouldn't be emitted at all. This event is not handled anywhere.

simplypixi replied:
fixed

simplypixi commented on frontend/globals/unity-util.ts:
```diff
@@ -729,8 +729,8 @@ export class UnityUtil {
 		UnityUtil.onLoaded();
 		UnityUtil.toUnity('LoadModel', UnityUtil.LoadingState.VIEWER_READY, JSON.stringify(params));
 
+		onStart();
```
I used it this approach because of relying on `onLoading` promise occurs some errors. I'll try to do this again tomorrow. Maybe I missed something.

carmenfan replied:
Try it again, let me know (or maybe commit that version) if it's not working (and how).

This promise is designed for exactly that and I'm worried we're bolting on a workaround for something to compensate for a bug that needs to be fixed.

simplypixi commented on frontend/globals/viewer.ts:
```diff
@@ -508,7 +513,14 @@ export class Viewer {
 		this.emit(Viewer.EVENT.START_LOADING);
 
-		UnityUtil.loadModel(this.account, this.model, this.branch, this.revision);
+		UnityUtil.loadModel({
```
1. Object is more flexible.
2. Which value do you mean?

carmenfan replied:
1. Why does it need to be more flexible? You're depicting the exact values of the parameters for this function call anyway? I'm not a fan of wrapping object around parameters unless justified. Its makes things more vague and bypasses build checks.

2. I mean the loadModel returns a promise, this is basically related to the comment on the unity-utils.ts bit :)

simplypixi replied:
1. At least because of few reasons:
- Forcing naming convention
- Non-ordering of specific arguments
- Shorter function's head

These reasons are probably not so important for you to make the change like this. I'll change it. ;)
Just because I'm curious. How does it bypass build checks?

2. Ok!

simplypixi commented:
> * [x]  pins should appear whilst the model is loading, not when it's fully loaded. ([loadModel in unityUtils](https://github.com/3drepo/3drepo.io/blob/master/frontend/globals/unity-util.ts#L722) returns a promise that is resolved when the model has started loading. This is the promise we want to wait on. I'm pretty sure it's exposed on the viewService somehow)
> * [x]  Same with viewpoint when the user enters a url with risk ID - it should wait on model loading, not model loaded
> * [x]  Resolve risks pins are being shown on load (pin on the right should not appear initially):
>   ![1](https://user-images.githubusercontent.com/11945337/52409796-710fc800-2ace-11e9-954b-0d370ec55835.gif)
> * [x]  I don't think the angular version does it, but I think the pins should update as the filter change (i.e. hidden issues needs to hide pin, shown issues need to show pins).
> * [x]  Only users who matches the risk owner job, admins or matches risk creator job can edit the risk. It should be viewable but all editing elements shouuld be disabled for anyone else. At the moment, it looks like anyone can edit them, when saving, the backend bounce back with an error and the front end doesn't really respond:
>   ![1](https://user-images.githubusercontent.com/11945337/52410827-93571500-2ad1-11e9-9ab4-73aa14646b07.gif)

@dleszcz Did you check that, just like I asked you for?

It's related with:
> I don't think this approach is reliable. This modelLoadingStarted would not be in the right state for the 2nd model load (i.e. I loaded a model, I went back to the teamspace view and selected another model).

dleszcz commented:
> > * Risk pins are hiddened when they're filtered out but they do not appear again when you cancel the filtering:
> 
> @dleszcz Did you check that, just like I asked you for?
Yes, I've checked this case carefully and it's working fine.

@carmenfan Referring to the problems with scroll. What system/browser you used?

carmenfan commented:
> > > * Risk pins are hiddened when they're filtered out but they do not appear again when you cancel the filtering:
> > 
> > 
> > @dleszcz Did you check that, just like I asked you for?
> 
> Yes, I've checked this case carefully and it's working fine.
> 
> @carmenfan Referring to the problems with scroll. What system/browser you used?

 win10/chrome ( 71.0.3578.98)

dleszcz commented:
> In the past, clicking on a pin would also open the Issues or Safetibase panel if not already open. It looks like this functionality was lost and it is also missing from the version on staging.

@Charence So, should we do this functionality?

carmenfan commented:
> > In the past, clicking on a pin would also open the Issues or Safetibase panel if not already open. It looks like this functionality was lost and it is also missing from the version on staging.
> 
> @Charence So, should we do this functionality?

That would be great yes

simplypixi commented:
@dleszcz I've change destination branch to staging. So, we have to rebase this branch. ;)

simplypixi commented on frontend/components/viewer/js/viewer.service.ts:
```diff
@@ -65,6 +66,7 @@ export class ViewerService {
 		};
 
 		this.initialised = $q.defer();
+		this.modelLoadingStarted = $q.defer();
```
Fixed

simplypixi commented:
> > > In the past, clicking on a pin would also open the Issues or Safetibase panel if not already open. It looks like this functionality was lost and it is also missing from the version on staging.
> > 
> > 
> > @Charence So, should we do this functionality?
> 
> That would be great yes

Done ;)

simplypixi commented:
@carmenfan I've fixed issues. Now, I need to rebase this branch but the conflicts appeared. We didn't change anything on backend side, so... Could help me with resolving the conflict in `integrated/teamspace.js`? Which version of the code is proper?
![image](https://user-images.githubusercontent.com/5683490/53026801-efbc1c00-3463-11e9-9ef6-983a5c508ab7.png)

sanmont3drepo commented:
> @carmenfan I've fixed issues. Now, I need to rebase this branch but the conflicts appeared. We didn't change anything on backend side, so... Could help me with resolving the conflict in `integrated/teamspace.js`? Which version of the code is proper?
> ![image](https://user-images.githubusercontent.com/5683490/53026801-efbc1c00-3463-11e9-9ef6-983a5c508ab7.png)

The proper one is the one that says "staging"

simplypixi commented:
@sanmont3drepo @carmenfan Rebased and ready for review ;)

simplypixi commented:
> @sanmont3drepo @carmenfan Rebased and ready for review ;)

@sanmont3drepo Backend tests fail. Could you check what is wrong? I think it's related to the conflict in `integrated/teamspace`.

Charence commented:
> @sanmont3drepo Backend tests fail. Could you check what is wrong? I think it's related to the conflict in `integrated/teamspace`.

@simplypixi 
It looks like `backend/test/integrated/teamspace.js` is missing an import.

Not sure how it was lost, but you can see what it should look like from this previous commit: https://github.com/3drepo/3drepo.io/commit/e25132906c71d42dc2c70d8bcb95b88ba56de4ca
